File: spring-data/src/main/java/dev/cerbos/queryplan/springdata/AttributeMapping.java
/*
 * Copyright 2021-2026 Zenauth Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

package dev.cerbos.queryplan.springdata;

import java.util.Map;
import java.util.Objects;

/**
 * Maps one plan variable (e.g. {@code request.resource.attr.ownerId} or
 * {@code request.resource.id}) to a scalar {@linkplain Field path} or a collection
 * {@linkplain Relation relation} on the JPA model. The map passed to
 * {@link SpringDataQueryPlanAdapter#toSpecification(dev.cerbos.sdk.PlanResourcesResult, Map)
 * toSpecification} is keyed by the full variable name:
 *
 * <pre>{@code
 * Map<String, AttributeMapping> MAPPING = Map.of(
 *     "request.resource.attr.ownerId",    AttributeMapping.field("owner.id"),
 *     "request.resource.attr.department", AttributeMapping.field("department"),
 *     "request.resource.attr.tags",       AttributeMapping.relation("tags", Map.of(
 *         "name", AttributeMapping.field("name"))));
 * }</pre>
 *
 * <p>Choosing a helper:
 *
 * <table border="1">
 *   <caption>Mapping helpers</caption>
 *   <tr><th>Helper</th><th>Use for</th></tr>
 *   <tr><td>{@link #field(String) field("aPath")}</td>
 *       <td>Simple column or {@code @Embedded} dotted path</td></tr>
 *   <tr><td>{@link #relation(String) relation("tags")}</td>
 *       <td>{@code @ElementCollection<String>} (bare values)</td></tr>
 *   <tr><td>{@link #relation(String, String) relation("tags", "name")}</td>
 *       <td>{@code @OneToMany} collection whose default member field is {@code name}</td></tr>
 *   <tr><td>{@link #relation(String, Map) relation("tags", Map.of("name", field("name")))}</td>
 *       <td>{@code @OneToMany<Tag>} with explicit nested field mapping</td></tr>
 *   <tr><td>{@link #relation(String, String, Map) relation("tags", "name", Map.of(...))}</td>
 *       <td>Both: a default member field for bare-value operators plus nested mappings for
 *           lambda bodies</td></tr>
 *   <tr><td>{@link Relation#withPositionField relation(...).withPositionField("position")}</td>
 *       <td>Any of the relations above, with the member field holding each element's list
 *           index, for positional reads such as {@code R.attr.tags[0]}</td></tr>
 * </table>
 *
 * <p>A variable that is missing from the map, or mapped in a way the operator cannot use,
 * throws {@link UnmappedAttributeException}.
 */
public sealed interface AttributeMapping permits AttributeMapping.Field, AttributeMapping.Relation {

    /**
     * Maps an attribute to a scalar JPA path on the entity.
     *
     * <p>A dotted path can go through {@code @Embedded} objects and to-one associations, e.g.
     * {@code field("details.pixelWidth")} or {@code field("owner.id")}. Associations are
     * LEFT-joined, so a missing one makes only its own comparison UNKNOWN.
     *
     * @param jpaPath entity property name, or a dot-separated path
     * @return the scalar mapping
     */
    static Field field(String jpaPath) {
        return new Field(jpaPath, null);
    }

    /**
     * Maps an attribute to a scalar JPA path whose column can be NULL, and declares how the
     * caller sends that NULL to {@code check()}. It overrides the call-level
     * {@link NullAttributeRepresentation}, so one call can mix both conventions.
     *
     * <p>Under {@link NullAttributeRepresentation#EXPLICIT}, {@code eq}, {@code ne} and
     * {@code in} are rendered so a NULL column gives a definite result, as CEL does, instead of
     * SQL UNKNOWN. Under {@link NullAttributeRepresentation#OMITTED}, {@code eq} and {@code ne}
     * against a bare null are UNKNOWN for a NULL column under both polarities, and other null
     * operands against this attribute are rejected.
     *
     * @param jpaPath entity property name, or a dot-separated path
     * @param nullAttributeRepresentation how a NULL in this column reaches {@code check()}
     * @return the scalar mapping
     */
    static Field field(String jpaPath, NullAttributeRepresentation nullAttributeRepresentation) {
        return new Field(jpaPath, Objects.requireNonNull(
                nullAttributeRepresentation, "nullAttributeRepresentation"));
    }

    /**
     * Maps an attribute to a collection whose elements are the compared values, typically an
     * {@code @ElementCollection} of strings.
     *
     * @param joinAttribute the entity's collection property name
     * @return the relation mapping
     */
    static Relation relation(String joinAttribute) {
        return new Relation(joinAttribute, null, Map.of());
    }

    /**
     * Maps an attribute to an entity collection whose {@code defaultMemberField} is used
     * wherever the policy treats elements as bare values. With {@code relation("tags", "name")},
     * {@code "urgent" in R.attr.tags} compares against {@code tag.name}.
     *
     * @param joinAttribute the entity's collection property name
     * @param defaultMemberField member-entity field used when the policy addresses the
     *        element as a bare value
     * @return the relation mapping
     */
    static Relation relation(String joinAttribute, String defaultMemberField) {
        return new Relation(joinAttribute, defaultMemberField, Map.of());
    }

    /**
     * Maps an attribute to an entity collection with mappings for the member fields that lambda
     * bodies reference. With {@code relation("tags", Map.of("name", field("name")))},
     * {@code t.name} in {@code R.attr.tags.exists(t, t.name == "x")} resolves on the member
     * entity. A nested entry may itself be a {@code relation(...)}, for multi-hop chains.
     *
     * @param joinAttribute the entity's collection property name
     * @param fields policy-facing member field name → mapping on the member entity
     * @return the relation mapping
     */
    static Relation relation(String joinAttribute, Map<String, AttributeMapping> fields) {
        return new Relation(joinAttribute, null, fields);
    }

    /**
     * Maps an attribute to an entity collection with both a {@linkplain #relation(String,
     * String) default member field} and {@linkplain #relation(String, Map) nested field
     * mappings}.
     *
     * @param joinAttribute the entity's collection property name
     * @param defaultMemberField member-entity field used when the policy addresses the
     *        element as a bare value
     * @param fields policy-facing member field name → mapping on the member entity
     * @return the relation mapping
     */
    static Relation relation(String joinAttribute, String defaultMemberField, Map<String, AttributeMapping> fields) {
        return new Relation(joinAttribute, defaultMemberField, fields);
    }

    /**
     * Scalar mapping. {@code nullAttributeRepresentation} is null when the mapping declares no
     * NULL convention. Create via {@link #field(String)} or
     * {@link #field(String, NullAttributeRepresentation)}.
     */
    record Field(String jpaPath, NullAttributeRepresentation nullAttributeRepresentation)
            implements AttributeMapping {
        public Field {
            Objects.requireNonNull(jpaPath, "jpaPath");
        }

        /** A mapping that declares no NULL convention; see {@link #field(String)}. */
        public Field(String jpaPath) {
            this(jpaPath, null);
        }
    }

    /**
     * Collection mapping. {@code defaultMemberField} and {@code positionField} may be null;
     * {@code fields} is copied. Create via the {@code relation(...)} factory methods, and
     * {@link #withPositionField} to declare the element order.
     */
    record Relation(String joinAttribute, String defaultMemberField,
                    Map<String, AttributeMapping> fields, String positionField)
            implements AttributeMapping {
        public Relation {
            Objects.requireNonNull(joinAttribute, "joinAttribute");
            Objects.requireNonNull(fields, "fields");
            fields = Map.copyOf(fields);
        }

        /** A relation with no declared element order. */
        public Relation(String joinAttribute, String defaultMemberField,
                        Map<String, AttributeMapping> fields) {
            this(joinAttribute, defaultMemberField, fields, null);
        }

        /**
         * Declares the member-entity field holding each element's zero-based position in the
         * list the application sends to {@code check()}, so a positional read such as
         * {@code R.attr.tags[0] == "x"} can be translated. Without it, a relation has no order a
         * plan can name and positional reads throw {@link UnsupportedPlanShapeException}.
         *
         * <p>The field must hold exactly the list index: {@code 0} for the first element, one
         * row per position, no gaps. A read past the last position is a CEL error, which
         * denies.
         *
         * @param positionField member-entity field holding the element's list index
         * @return the relation mapping with its order declared
         */
        public Relation withPositionField(String positionField) {
            return new Relation(joinAttribute, defaultMemberField, fields,
                    Objects.requireNonNull(positionField, "positionField"));
        }
    }
}

File: activerecord/lib/cerbos/active_record/attribute_mapping.rb
# frozen_string_literal: true

require_relative "errors"

module Cerbos
  module ActiveRecord
    # Maps a plan `variable` name (e.g. `"request.resource.attr.ownerId"`) to the model, as a
    # scalar {Field} or a collection {Relation}.
    #
    #     MAPPING = {
    #       "request.resource.attr.ownerId"    => Cerbos::ActiveRecord.field("owner_id"),
    #       "request.resource.attr.department" => Cerbos::ActiveRecord.field("owner.department"),
    #       "request.resource.attr.tags"       => Cerbos::ActiveRecord.relation(
    #         :tags, member_field: "name", fields: {"name" => Cerbos::ActiveRecord.field("name")}
    #       )
    #     }
    #
    # An unmapped variable, or a mapping its operator cannot use, raises
    # {UnmappedAttributeError}. The adapter never guesses a column.
    module AttributeMapping
      # How the caller sends a NULL column to Cerbos. See {Cerbos::ActiveRecord.field}.
      NULL_REPRESENTATIONS = %i[explicit omitted].freeze

      # A scalar mapping.
      #
      # `path` is a column, or a dotted path through `belongs_to` / `has_one` associations
      # (e.g. `"owner.department"`). A path becomes a correlated scalar subquery, so it cannot
      # add rows.
      #
      # `null_representation` overrides the call's `null_attribute_representation` for this
      # attribute; `nil` keeps the call's value.
      #
      # Make one with {Cerbos::ActiveRecord.field}.
      #
      # @attr path [String] the column, or a dotted path through to-one associations.
      # @attr null_representation [Symbol, nil] `:explicit`, `:omitted`, or `nil` for the
      #   convention of the call.
      Field = Struct.new(:path, :null_representation) do
        # Specify a scalar mapping.
        #
        # @param path [String, Symbol] the column, or a dotted path through to-one associations.
        # @param null_representation [Symbol, String, nil] `:explicit`, `:omitted`, or `nil`.
        #
        # @raise [ArgumentError] when `path` is `nil` or `null_representation` is invalid.
        def initialize(path:, null_representation: nil)
          raise ArgumentError, "path is required" if path.nil?

          unless null_representation.nil? ||
              NULL_REPRESENTATIONS.include?(null_representation.to_sym)
            raise ArgumentError,
              "null_representation must be :explicit or :omitted, got " \
              "#{null_representation.inspect}"
          end

          super(path: path.to_s, null_representation: null_representation&.to_sym)
        end

        # @return [Array<String>] the path in parts: the association hops and then the column.
        def segments
          @segments ||= path.split(".").freeze
        end
      end

      # A collection mapping.
      #
      # `association` is a collection association on the owning model. `through:` chains
      # become joins in one correlated subquery. Map a `has_one` as a dotted {Field} path.
      #
      # `member_field` is the column used when the collection is a list of plain values, so
      # `"urgent" in R.attr.tags` compares `tag.name`, not the `Tag` record.
      #
      # `fields` maps member names used in lambda bodies, e.g. `"name"` for
      # `R.attr.tags.exists(t, t.name == "x")`. An entry can be a relation, for multi-hop
      # chains like `R.attr.categories.exists(c, c.subCategories.exists(s, ...))`.
      #
      # Make one with {Cerbos::ActiveRecord.relation}.
      #
      # @attr association [Symbol] the association name.
      # @attr member_field [String, nil] the column for a bare element value.
      # @attr fields [Hash{String => Field, Relation}] the member mappings for lambda bodies.
      Relation = Struct.new(:association, :member_field, :fields) do
        # Specify a collection mapping.
        #
        # @param association [String, Symbol] the association name.
        # @param member_field [String, Symbol, nil] the column for a bare element value.
        # @param fields [Hash{String, Symbol => Field, Relation}] the member mappings for lambda
        #   bodies.
        #
        # @raise [ArgumentError] when `association` is `nil` or a `fields` value is not a mapping.
        def initialize(association:, member_field: nil, fields: {})
          raise ArgumentError, "association is required" if association.nil?

          fields.each do |name, mapping|
            unless mapping.is_a?(Field) || mapping.is_a?(Relation)
              raise ArgumentError, "fields[#{name.inspect}] must be a field or relation mapping"
            end
          end

          super(
            association: association.to_sym,
            member_field: member_field&.to_s,
            fields: fields.transform_keys(&:to_s).freeze
          )
        end
      end
    end

    # Makes a scalar {AttributeMapping::Field} mapping.
    #
    # @param path [String, Symbol] a column, or a dotted path through to-one associations.
    # @param null_representation [Symbol, nil] how a NULL column is sent: `:explicit` (as a
    #   null value) or `:omitted` (not sent). `nil` (default) uses the call's value. Declaring
    #   it makes `eq`, `ne` and `in` match the PDP on NULL rows
    #   (cerbos/query-plan-adapters#308).
    #
    # @return [AttributeMapping::Field]
    #
    # @raise [ArgumentError] when `path` is `nil` or `null_representation` is invalid.
    #
    # @example A column on the model
    #   Cerbos::ActiveRecord.field("owner_id")
    #
    # @example A column on a to-one parent, sent as an explicit null
    #   Cerbos::ActiveRecord.field("owner.department", null_representation: :explicit)
    def self.field(path, null_representation: nil)
      AttributeMapping::Field.new(path: path, null_representation: null_representation)
    end

    # Makes a collection {AttributeMapping::Relation} mapping.
    #
    # @param association [String, Symbol] the association name on the model that owns it.
    # @param member_field [String, Symbol, nil] the column for a bare element value.
    # @param fields [Hash{String => AttributeMapping::Field, AttributeMapping::Relation}]
    #   the member mappings for lambda bodies.
    #
    # @return [AttributeMapping::Relation]
    #
    # @raise [ArgumentError] when `association` is `nil` or a `fields` value is not a mapping.
    #
    # @example Tags compared by name
    #   Cerbos::ActiveRecord.relation(
    #     :tags, member_field: "name", fields: {"name" => Cerbos::ActiveRecord.field("name")}
    #   )
    #
    # @example A chain through a parent
    #   Cerbos::ActiveRecord.relation(:categories, fields: {
    #     "subCategories" => Cerbos::ActiveRecord.relation(:sub_categories, fields: {
    #       "name" => Cerbos::ActiveRecord.field("name")
    #     })
    #   })
    def self.relation(association, member_field: nil, fields: {})
      AttributeMapping::Relation.new(association: association, member_field: member_field, fields: fields)
    end
  end
end
